src/frontend/src/utils/connected_components.rs

```diff
@@ -0,0 +1,125 @@
+// Copyright 2022 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::{HashMap, HashSet};
+
+#[derive(Debug)]
+pub(crate) struct ConnectedComponentLabeller {
+    vertex_to_label: HashMap<usize, usize>,
```

Comment:
Why do we have a `label` here? Can we directly operate on vertex id?

Reply:
We could indeed have `vertex_to_connected_vertices`, and `vertex_to_connected_edges`, that maps to Arc/Rc, but I didn't want to focus too much on optimizing the performance, its only meant for small graphs...

Reply:
Basically I acknowledge its not necessarily the most efficient impl but I didn't want to spend too much time on optimizing it... if someone wants to improve it you may submit a PR :)

Reply:
I added a comment that it could be optimized further (maybe I will do it in next PR...)